File: static/solutions/cpp/java/src/main/java/com/epi/BinaryTreePreorderTraversalIterative.java
package com.epi;

import com.epi.BinaryTreePrototypeTemplate.BinaryTreeNode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

import static com.epi.BinaryTreeUtils.generatePreorder;

public class BinaryTreePreorderTraversalIterative {
  // @include
  public static List<Integer> preorderTraversal(BinaryTreeNode<Integer> tree) {
    Deque<BinaryTreeNode<Integer>> path = new LinkedList<>();
    path.addFirst(tree);
    List<Integer> result = new ArrayList<>();
    while (!path.isEmpty()) {
      BinaryTreeNode<Integer> curr = path.removeFirst();
      if (curr != null) {
        result.add(curr.data);
        path.addFirst(curr.right);
        path.addFirst(curr.left);
      }
    }
    return result;
  }
  // @exclude

  public static void main(String[] args) {
    // 3
    // 2 5
    // 1 4 6
    BinaryTreeNode<Integer> tree = new BinaryTreeNode<>(3);
    tree.left = new BinaryTreeNode<>(2);
    tree.left.left = new BinaryTreeNode<>(1);
    tree.right = new BinaryTreeNode<>(5);
    tree.right.left = new BinaryTreeNode<>(4);
    tree.right.right = new BinaryTreeNode<>(6);
    List<Integer> res = preorderTraversal(tree);
    List<Integer> goldenRes = generatePreorder(tree);
    assert(res.equals(goldenRes));
  }
}

File: static/cpp/binary-tree-preorder-traversal-iterative.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <memory>
#include <stack>
#include <vector>

#include "./Binary_tree_prototype.h"
#include "./Binary_tree_utils.h"

using std::cout;
using std::endl;
using std::equal;
using std::make_unique;
using std::stack;
using std::unique_ptr;
using std::vector;

// @include
vector<int> PreorderTraversal(const unique_ptr<BinaryTreeNode<int>>& tree) {
  stack<BinaryTreeNode<int>*> path;
  path.emplace(tree.get());
  vector<int> result;
  while (!path.empty()) {
    auto curr = path.top();
    path.pop();
    if (curr != nullptr) {
      result.emplace_back(curr->data);
      path.emplace(curr->right.get());
      path.emplace(curr->left.get());
    }
  }
  return result;
}
// @exclude

int main(int argc, char** argv) {
  //      3
  //    2   5
  //  1    4 6
  unique_ptr<BinaryTreeNode<int>> tree = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{3, nullptr, nullptr});
  tree->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{2, nullptr, nullptr});
  tree->left->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{1, nullptr, nullptr});
  tree->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{5, nullptr, nullptr});
  tree->right->left = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{4, nullptr, nullptr});
  tree->right->right = make_unique<BinaryTreeNode<int>>(
      BinaryTreeNode<int>{6, nullptr, nullptr});
  auto res = PreorderTraversal(tree), golden_res = generate_preorder(tree);
  assert(equal(res.cbegin(), res.cend(), golden_res.cbegin(),
               golden_res.cend()));
  return 0;
}
